### ISSUE TITLE
Is the validation set used directly during training ? 

### ISSUE THREAD

**Issue by ervgan** [ultralytics/yolov5#10822]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello,

I have noticed that when training my dataset with train.py, the code also scans my val dataset: 
val: Scanning /content/drive/MyDrive/Colab_Notebooks/Datasets/labels/val.cache... 61 images, 0 backgrounds, 0 corrupt: 100% 61/61 [00:00<?, ?it/s]

so is the validation set already used during training to fine tune hyper parameters ? if so, what's the point of val.py ?

Many thanks !

**Comment by JustasBart** [comment]
Hi, I believe it calls val.py at the end of every epoch for you, but you can also do it manually later on if you want to get the stats of your model.

And also yes, of course it uses the validation set directly during training!

Good luck! :rocket:

**Comment by EdjeElectronics** [comment]
Per issue #6023 , the validation set is not actually used to adjust hyperparameters during training. It is only used to compute metrics at the end of each epoch. The validation set does not have any impact on the model's weights.

**Comment by glenn-jocher** [comment]
@EdjeElectronics thanks for your keen observation! You're correct; the validation set is not utilized to fine-tune hyperparameters during training. It's primarily employed to calculate metrics at the end of each epoch, with no direct impact on the model's weights. If you have any further questions, feel free to ask!

**Comment by tomhoq** [comment]
@glenn-jocher  Wouldn't it be better to have a val and a test set instead of just a val set? Isn't the entire purpose of the validation set to prevent overfitting to the training data by updating the weights?

**Comment by glenn-jocher** [comment]
Hi @tomhoq,

Thank you for your insightful question! Let's clarify the roles of the validation and test sets in the context of training machine learning models, specifically with YOLOv5.

### Validation Set vs. Test Set

1. **Validation Set**: This set is used during training to monitor the model's performance and compute metrics at the end of each epoch. It helps in tuning hyperparameters and provides an early indication of how well the model is generalizing to unseen data. However, it does not directly influence the model's weights.

2. **Test Set**: This set is used after the model has been fully trained to evaluate its performance on completely unseen data. It provides an unbiased evaluation of the final model's performance.

### Why Both Sets Are Important

- **Preventing Overfitting**: The validation set helps in detecting overfitting during training by providing a checkpoint to evaluate the model's performance. However, it does not update the weights directly. Instead, it helps in making decisions such as early stopping or hyperparameter adjustments.
  
- **Final Evaluation**: The test set is crucial for assessing the model's true performance. Since it is not used during training or validation, it offers a final, unbiased evaluation of the model's ability to generalize.

### Implementation in YOLOv5

In YOLOv5, the validation set is used to compute metrics like mAP (mean Average Precision) at the end of each epoch, which helps in monitoring the training process. If you want to include a test set for final evaluation, you can split your dataset accordingly and use the `val.py` script to evaluate your model on the test set after training.

Here's a brief example of how you might structure your dataset:

```plaintext
dataset/
├── train/
│   ├── images/
│   └── labels/
├── val/
│   ├── images/
│   └── labels/
└── test/
    ├── images/
    └── labels/
```

After training, you can run:

```bash
python val.py --data dataset.yaml --weights best.pt --task test
```

This will evaluate your trained model on the test set.

### Next Steps

If you haven't already, please ensure you are using the latest versions of `torch` and YOLOv5 from [our repository](https://github.com/ultralytics/yolov5). If you encounter any issues, providing a minimum reproducible example will help us investigate further. You can find more details on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Thank you for your engagement and contributions to the YOLO community! If you have any more questions or need further assistance, feel free to ask.

**Comment by tomhoq** [comment]
@glenn-jocher thank you for the quick answer!

So the model won't learn from the validation set correct? This means that, in theory, evaluating the performance with the validation set or with the testing set, imagining they have similar distributions, would reward similar results?

**Comment by glenn-jocher** [comment]
Hi @tomhoq,

You're absolutely right! The model does not learn from the validation set. Instead, the validation set is used to monitor the model's performance during training by computing metrics at the end of each epoch. This helps in detecting issues like overfitting and underfitting without influencing the model's weights.

### Validation Set vs. Test Set

- **Validation Set**: Used during training to provide feedback on the model's performance. It helps in tuning hyperparameters and making decisions such as early stopping. However, it does not directly affect the training process or update the model's weights.

- **Test Set**: Used after the model has been fully trained to evaluate its performance on completely unseen data. It provides an unbiased assessment of the model's generalization capabilities.

### Similar Distributions

If the validation and test sets have similar distributions, you can expect the performance metrics (such as accuracy, precision, recall, and mAP) to be similar. However, it's always a good practice to evaluate your model on a separate test set to ensure that the performance metrics are not biased by any specific characteristics of the validation set.

### Example Workflow

1. **Training with Validation Set**:
    ```bash
    python train.py --img 640 --epochs 50 --data dataset.yaml --weights yolov5s.pt
    ```

2. **Evaluating with Test Set**:
    ```bash
    python val.py --data dataset.yaml --weights best.pt --task test
    ```

This way, you ensure that your model's performance is robust and generalizes well to new, unseen data.

### Next Steps

If you encounter any issues or have further questions, please ensure you are using the latest versions of `torch` and YOLOv5 from [our repository](https://github.com/ultralytics/yolov5). If the issue persists, providing a minimum reproducible example will help us investigate further. You can find more details on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Thank you for your engagement and contributions to the YOLO community! If you have any more questions or need further assistance, feel free to ask. 😊

**Comment by MarioPasc** [comment]
Hello,

I just had a quick question. The model will always perform better on the training set than in the validation set, but, for visualization and checking purposes, I'm interested in comparing this performance. The problem is that, when in the training parameters we set `val = False`, it does not perform inference on the training set rather than the validation set and store the metrics, instead, it deactivates the validation function completely. 

I've seen that the validation task has an argument `split`, which can be set to `train`. I believe this functionality is the one that I'd like to have: measuring the mAP@50-95 of my model, while training, on the training set. The idea is to set 2 trainings at the same time for YOLOv8l (`deterministic = True`, `seed = 0`, same dataset), one will perform validation on the val set, and the other on the train, and then compare them. 

This attached plot is what i got for that exact set up I previously described. Two trainings at the same time, one had `val=True`, other had `val=False`. If I'm missing something, I'd appreciate the feedback :) 

Thank you!

![map_evolution_plot](https://github.com/user-attachments/assets/5af6380d-b496-455d-818b-82ebeb2b5f79)

**Comment by pderrenger** [comment]
Hello,

To evaluate your model on the training set during training, you can indeed use the `split` argument in the validation task to set it to `train`. This will allow you to compute metrics like mAP on the training set. Ensure your setup is consistent across both runs for a fair comparison. If you encounter any issues, please verify with the latest YOLOv5 version. Let us know if you need further assistance!
